magda-sleuther-broken-link/src/index.ts

```diff
@@ -0,0 +1,370 @@
+import Registry from "@magda/typescript-common/dist/Registry";
+import { Record } from "@magda/typescript-common/dist/generated/registry/api";
+import unionToThrowable from "@magda/typescript-common/dist/util/union-to-throwable";
+import sleuther from "@magda/sleuther-framework/dist/index";
+import * as _ from "lodash";
+import * as request from "request";
+import * as http from "http";
+import * as Client from "ftp";
+import * as LRU from "lru-cache";
+import * as URI from "urijs";
+import retryBackoff from "@magda/typescript-common/dist/retryBackoff";
+
+const brokenLinkAspectDefinition = {
+  id: "source-link-status",
+  name: "Details about the downloadURL link status of a distribution",
+  jsonSchema: require("@magda/registry-aspects/source-link-status.schema.json")
+};
+const datasetQualityAspectDef = {
+  id: "dataset-quality-rating",
+  name: "Data Quality Rating",
+  jsonSchema: require("@magda/registry-aspects/dataset-quality-rating.schema.json")
+};
+const ID = "sleuther-broken-link";
+const host = process.env.HOST || ID;
+
+const registry = new Registry({
+  baseUrl:
+    process.env.REGISTRY_URL ||
+    process.env.npm_package_config_registryUrl ||
+    "http://localhost:6100/v0"
+});
+
+sleuthBrokenLinks(registry);
+
+function sleuthBrokenLinks(registry: Registry) {
+  async function onRecordFound(record: Record) {
+    const distributions: Record[] = record.aspects["dataset-distributions"]
+      ? record.aspects["dataset-distributions"].distributions
+      : [];
+
+    // Check each link
+    const linkChecks: Array<
+      [string, () => Promise<BrokenLinkSleuthingResult>]
+    > = distributions.map((distribution: Record) =>
+      checkDistributionLink(
+        distribution,
+        distribution.aspects["dcat-distribution-strings"]
+      )
+    );
+
+    // Group the checks against their host so that we're only making one request per site simultaneously.
+    const brokenLinkChecksByHost: Promise<BrokenLinkSleuthingResult[]>[] = _(
+      linkChecks
```

Comment:
This will only limit per dataset, unless datasets are processed serially by the sleuther framework, in which case the retry parameters may need to tweaked more (each broken link will take ~20 seconds of retries).